src/vs/editor/browser/viewParts/lines/viewLine.ts

```diff
@@ -184,7 +187,15 @@ export class ViewLine implements IVisibleLine {
 				const endColumn = (selection.endLineNumber === lineNumber ? selection.endColumn : lineData.maxColumn);
 
 				if (startColumn < endColumn) {
-					actualInlineDecorations.push(new LineDecoration(startColumn, endColumn, 'inline-selected-text', InlineDecorationType.Regular));
+					if (this._options.renderWhitespace !== 'selection') {
+						actualInlineDecorations.push(new LineDecoration(startColumn, endColumn, 'inline-selected-text', InlineDecorationType.Regular));
+					} else {
+						if (!selectionsOnLine) {
+							selectionsOnLine = [];
+						}
+
+						selectionsOnLine.push(new Range(lineNumber, startColumn, lineNumber, endColumn));
```

Comment:
Replace here with `LineRange` (defined in another comment). You can actually test and see how adding line numbers to the mix breaks the line repaint optimizations by doing something like the following:
- open file A in two editors, side by side.
- add logging to the rendering skip code below:
```
if (this._renderedViewLine && this._renderedViewLine.input.equals(renderLineInput)) {
	// no need to do anything, we have the same render input
	return false;
}
```
- in the first editor do a selection on line X
- in the second editor, insert a line above the line X such that line X in the first editor moves down. 
- Line X should not be rendered again just because it got moved down...